### ISSUE TITLE
Nested drawers issue

### ISSUE THREAD

**Issue by lorenzsell** [react-navigation/react-navigation#1261]
I have a drawer nested in a stack nav.  that drawer has a stack nav with another drawer nested inside it.  Before I upgraded versions, I had no trouble opening the main drawer when I was on the main screen with navigate('DrawerOpen') and opening the secondary drawer when I was on the nested stack nav.  After upgrading, navigate('DrawerOpen') always opens the sub drawer and I can't figure out how to open the top level drawer.

**Comment by lorenzsell** [comment]
@cooperka - sorry to ping you directly on this.  i know you worked on some drawer stuff before.  i've literally spent days trying to debug this and still struggling to make it work properly.  do you have any thoughts on what the problem might be?

**Comment by cooperka** [comment]
Hmm, nothing in particular stands out to me. It could either be a new bug, or an old bug you were accidentally relying on that got fixed ;)

To figure out which commit broke your code, you could try using [`git bisect`](http://webchick.net/node/99) which is a great tool for quickly discovering where an issue started. It might be easiest to use the react-navigation example app and tweak it in order to reproduce the bug. If you're able to figure out which commit caused it to stop working, please post it here and we can try to help further!

**Comment by lorenzsell** [comment]
@cooperka - I'm still struggling to figure out what's causing this.  I've copied my nav set up into a bare bones app.  Would you mind taking a look to see if anything strikes you as wrong?  As far as I can tell I've followed all the examples perfectly: https://github.com/lorenzsell/react-nav-test

**Comment by cooperka** [comment]
I appreciate the clean code, it looks great! Nothing I can tell that's wrong with it though :( have you tried `git bisect` yet? It's usually helpful to know which change broke your code, then you know what to look for.

You might also try using `getStateForAction(NavigationActions.navigate({ routeName: 'Login' }))` as suggested here: https://github.com/react-community/react-navigation/pull/1311. I'm not sure if it's related.

**Comment by lorenzsell** [comment]
ok.  i'm trying to figure this out with git bisect.  I'm somewhat new to git and npm.  maybe you can point me in the right direction.  if i understand correctly, in order to do this, i have to fork react-navigation and clone it locally.  then i have to link my project to the locally cloned version by npm installing from the local directory.  then i git bisect the local react-navigation version and refresh my simulator to see where the problem occurs.

I'm having trouble npm installing the local version.  I get:

`Error: Couldn't find preset "react-native" relative to directory "/Users/lorenzsell/DEV/react-navigation"`

Maybe you can tell me if I'm going about this all wrong before I spend hours trying to get it to work.  Thank you!

**Comment by lorenzsell** [comment]
@cooperka - I ended up doing a "manual" bisect and going through all the commits until I found the culprit: https://github.com/react-community/react-navigation/commit/2dc869fedd27d304c2f0ddd606761320b8bec923.  (Please feel free to weigh in on how I could have done that in a more efficient way).

It seems to be an issue with the following code:

     ```
// Handle navigation to other child routers that are not yet pushed
      if (action.type === NavigationActions.NAVIGATE) {
        const childRouterNames = Object.keys(childRouters);
        for (let i = 0; i < childRouterNames.length; i++) {
          const childRouterName = childRouterNames[i];
          const childRouter = childRouters[childRouterName];
          if (childRouter) {
            // For each child router, start with a blank state
            const initChildRoute = childRouter.getStateForAction(
              NavigationActions.init(),
            );
            // Then check to see if the router handles our navigate action
            const navigatedChildRoute = childRouter.getStateForAction(
              action,
              initChildRoute,
            );
            let routeToPush = null;
            if (navigatedChildRoute === null) {
              // Push the route if the router has 'handled' the action and returned null
              routeToPush = initChildRoute;
            } else if (navigatedChildRoute !== initChildRoute) {
              // Push the route if the state has changed in response to this navigation
              routeToPush = navigatedChildRoute;
            }
            if (routeToPush) {
              return StateUtils.push(state, {
                ...routeToPush,
                key: _getUuid(),
                routeName: childRouterName,
              });
            }
          }
        }
      }
```

I've updated:

`if (action.type === NavigationActions.NAVIGATE) {`

to

```
if (
        action.type === NavigationActions.NAVIGATE &&
        action.routeName !== 'DrawerOpen' &&
        action.routeName !== 'DrawerClose'
      ) {
```
in this pull request: #1379.

@grabbou - maybe you can weigh in, I think this commit was originally from you.  Is this a bug or intended behavior?  

Also, this is my first pull request ever.  So if there are any best practices I should be following, please enlighten me.  I love this project and I'm hoping to understand the world of open source contributions much better.  Thank you.

**Comment by cooperka** [comment]
Thanks for submitting the PR, looks fine! I'm not sure if it's intended behavior though, so I'll let others weigh in.

**Comment by boliveira** [comment]
+1

**Comment by boliveira** [comment]
The problem is in the TabRouter getStateForAction associated with the Drawer. 
It looks first for a route in the activeTab, that is in the case of Drawer, the DrawerClose route. As the DrawerClose route has the other routes for the other screens, it will eventually find the nested drawer, DrawerOpen route and selects it. 

Maybe another router should be used for it?

**Comment by kelset** [comment]
Pinging OP @lorenzsell since this issue is quite old and related to an old version of the lib. If no answer I'll close in 7 days.

That said, also the PR submitted a while back I think had a discussion about going another approach so not sure how to handle if still present.

**Comment by lorenzsell** [comment]
@kelset - the particular issue here is still an issue.  I've been using a version of the library with my simple fix that wasn't accepted as a pull request.  A different approach was suggested but since I didn't get more direction I wasn't sure where to go with that.  I'm happy to work on better approach but I'm somewhat new to Javascript and would value some feedback on the best way to address this issue.

But I did just check it against v1.0.0-beta.15 and it's still an issue as far as I can tell.  I have a top level drawer with a stack navigator inside it that has a drawer inside it.  When I try to open the top level drawer, the app jumps to the stack level drawer.

**Comment by brentvatne** [comment]
@ericvicenti - I think this might be solved in the refactor of `createNavigator` given that we're creating a separate router for Drawer

**Comment by brentvatne** [comment]
in 2.0 you can call `openDrawer` and `closeDrawer` directly on navigation prop: https://github.com/react-navigation/react-navigation/blob/ec1694f909f895b2be60c5d44b4fb5b278032aff/src/routers/DrawerRouter.js#L21-L24

notice also that they take a key. you can add a custom action on your top level drawer like this:

```js
let MyTopLevelDrawer = createDrawerNavigator({
  // routes
}, {
  getCustomActionCreators: (route, stateKey) => ({
    openTopLevelDrawer: () => DrawerActions.openDrawer({key: stateKey}),
    closeTopLevelDrawer: () => DrawerActions.closeDrawer({key: stateKey}),
  }),
})
```

then from a screen within both drawers you can just call `openDrawer` / `closeDrawer` for the inner drawer and `openTopLevelDrawer` / `closeTopLevelDrawer` (or whatever you want to call these

**Comment by brentvatne** [comment]
moving this to umbrella issue

**Comment by Xyzor** [comment]
@brentvatne I've tried what you suggested, but the `getCustomActionCreators` isn't get called.

**Comment by KamauIan** [comment]
> in 2.0 you can call `openDrawer` and `closeDrawer` directly on navigation prop:
> 
> https://github.com/react-navigation/react-navigation/blob/ec1694f909f895b2be60c5d44b4fb5b278032aff/src/routers/DrawerRouter.js#L21-L24
> 
> notice also that they take a key. you can add a custom action on your top level drawer like this:
> 
> ```js
> let MyTopLevelDrawer = createDrawerNavigator({
>   // routes
> }, {
>   getCustomActionCreators: (route, stateKey) => ({
>     openTopLevelDrawer: () => DrawerActions.openDrawer({key: stateKey}),
>     closeTopLevelDrawer: () => DrawerActions.closeDrawer({key: stateKey}),
>   }),
> })
> ```
> 
> then from a screen within both drawers you can just call `openDrawer` / `closeDrawer` for the inner drawer and `openTopLevelDrawer` / `closeTopLevelDrawer` (or whatever you want to call these)

getCustomActionCreators is no longer available in version 4.0.10. Is there another way or an equivalent to getCustomActionCreators to implement this problem? Thanks.
